internal/envoy/versions_test.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/func-e/internal/version"
+)
+
+func TestFuncEVersions_FindLatestPatch(t *testing.T) {
+	type testCase struct {
+		name     string
+		input    version.Version
+		versions map[version.Version]version.Release
+		want     version.Version
+	}
+
+	tests := []testCase{
+		{
+			name:  "zero",
+			input: "1.20",
+			versions: map[version.Version]version.Release{
+				"1.20.0": {},
+			},
+			want: "1.20.0",
+		},
+		{
+			name:  "upgradable",
+			input: "1.18",
+			versions: map[version.Version]version.Release{
+				"1.18.3": {},
+				"1.18.4": {},
+			},
+			want: "1.18.4",
+		},
+		{
+			name:  "notfound",
+			input: "1.1",
+			versions: map[version.Version]version.Release{
+				"1.20.0": {},
+			},
+			want: "",
+		},
+	}
+
+	ctx := context.Background()
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			tester := newFuncEVersionsTester(tc.versions)
+			have, err := tester.feV.FindLatestPatch(ctx, tc.input)
+			if tc.want == "" {
+				require.Error(t, fmt.Errorf("couldn't find latest version for %s", tc.input))
```

Comment:
I guess we could do something like this!
```suggestion
				require.Errorf(t, "couldn't find latest version for %s", tc.input)
```

Reply:
Cool. Thanks!